Feature/lit 261 update developer documentation for otp auth (#77)
LIT-Protocol/docs: docs/SDK/Explanation/authentication/sessionSigs/authMethods/email-sms.md

# Email / SMS

Email and SMS authentication provides users with a convenient way to verify their identity using one-time passwords (OTP) sent to their registered email address or phone number. 

Authenticating with OTP codes delivered via email or SMS is a two-step process. First, an OTP code is initiated and sent to the user's registered email or phone number. The user-provided code is then verified. Upon successful verification, a signed JSON Web Token (JWT) is generated. This token will be validated when creating session signatures.

:::note

> Codes sent to users via email will be recieved from `noreply@litprotocol.com`. Codes sent to users via sms will include `lit-verification` within the sms message.

Reviewer: ```suggestion
Codes sent to users via email will be recieved from `noreply@litprotocol.com`. Codes sent to users via SMS will include `lit-verification` within the SMS message.
```

:::
### Register user with email or SMS

```javascript
const authClient = new LitAuthClient({
    litRelayConfig: {
        relayApiKey: '<Your Lit Relay Server API Key>',
    }
});

// starting a validation session
let session = authClient.initProvider(ProviderType.Otp,{
            userId: '<User email or phone number>'
});

let status = await session.sendOtpCode();
let authMethod = await session.authenticate({

>     code: "user entered otp code"

Reviewer: ```suggestion
    code: "<User entered OTP code>"
```

If you are using Lit Relay Server, you will need to request an API key [here](https://forms.gle/RNZYtGYTY9BcD9MEA).

:::

### Minting via Contract

An alternative to minting the PKP NFT via the Lit Relay Server is to send a transaction to the smart contract yourself. You can reference the following example data that is passed to the `mintNextAndAddAuthMethods` method of the `PKPHelper` smart contract:

- `keyType` is `2`
- `permittedAuthMethodTypes` is `[7]`
- `permittedAuthMethodIds` is an array with 1 element being the user's email or phone number.
- `permittedAuthMethodScopes` is an array with 1 zero-initialized element, e.g. `[[ethers.BigNumber.from("0")]]`
- `addPkpEthAddressAsPermittedAddress` is `true`
- `sendPkpToItself` is `true`

### Authenticating to Fetch PKP information

```javascript
const authClient = new LitAuthClient({
    litRelayConfig: {
        relayApiKey: '<Your Lit Relay Server API Key>',
    }
});

// starting a validation session
let session = authClient.initProvider(ProviderType.Otp,{
            userId: '<User email or phone number>'
});

let status = await session.sendOtpCode();
let authMethod = await session.authenticate({

>     code: "<User entered otp code>"

Reviewer: ```suggestion
    code: "<User entered OTP code>"
```

# Email / SMS

> When authenticating with email or sms provided otp codes, it is a two step process. First initating a code be sent to a email / phone number, then asserting on the user provided code which will generated a signed Json Web Token if successful. This token will be validated when signing Session Signatures as the `accessToken` within the returned `Auth Method`.

Reviewer: ```suggestion
Email and SMS authentication provides users with a convenient way to verify their identity using one-time passwords (OTP) sent to their registered email address or phone number. 

Authenticating with OTP codes delivered via email or SMS is a two-step process. First, an OTP code is initiated and sent to the user's registered email or phone number. The user-provided code is then verified. Upon successful verification, a signed JSON Web Token (JWT) is generated. This token will be validated when creating session signatures.
```

:::note
Codes sent to users via email will be recieved from `noreply@litprotocol.com`. Codes sent to users via sms will include `lit-verification` within the sms message.
:::

> ### Register user with sms / email auth

Reviewer: ```suggestion
### Register user with email or SMS
```

```javascript
const authClient = new LitAuthClient({
    litRelayConfig: {

>         relayApiKey: 'your-api-key',

Reviewer: for consistent styling and making these filler values more distinct,

```suggestion
        relayApiKey: '<Your Lit Relay Server API Key>',
```

check out https://github.com/LIT-Protocol/docs/blob/main/docs/SDK/Explanation/authentication/sessionSigs/authMethods/socialLogin.md as an example

    "authMethodType": 7
}
```

:::note 
 Tokens expire after 30 minutes, and must be validated for session signature signing within that period
:::

## Generating `SessionSigs`

After successfully authenticating with an `AuthMethod`, you can generate `Session Signatures` using the provider's `getSessionSigs` method. The `getSessionSigs` method takes in an `AuthMethod` object, a PKP public key, and other session-specific arguments such as `resourceAbilityRequests` and returns a `SessionSig` object.

```javascript
// Get session signatures for the given PKP public key and auth method
const sessionSigs = await provider.getSessionSigs({
  authMethod: '<AuthMethod object returned from authenticate()>',
  sessionSigsParams: {
    chain: 'ethereum',
    resourceAbilityRequests: [{
      resource: litResource,
      ability: LitAbility.AccessControlConditionDecryption
    }],
  },
});
```

> # Technical Details

Reviewer: i think this section is only really required if we provide details that aren't entirely mentioned above (see https://developer.litprotocol.com/SDK/Explanation/authentication/sessionSigs/authMethods/webAuthn#technical-details as an example)
